python/segmentation/models/model_pspnet.py

```diff
@@ -0,0 +1,119 @@
+                )
+    def forward(self, x, img_size):
+        return F.interpolate(self.aux(x), img_size, mode='bilinear', align_corners=False)
+        
+        
+class pyramid_pooling_module(nn.Module):
+    def __init__(self, in_channels, out_channels, bin_sizes):
+        super(pyramid_pooling_module, self).__init__()
+        
+        # create pyramid pooling layers for each level
+        self.pyramid_pool_layers = []
+        for bin_sz in bin_sizes:
+            self.pyramid_pool_layers.append(nn.Sequential(
+                nn.AdaptiveAvgPool2d(bin_sz),
+                nn.Conv2d(in_channels, out_channels, kernel_size=1, bias=False),
+                nn.BatchNorm2d(out_channels),
+                nn.ReLU(inplace=True)
+            ))
+        self.pyramid_pool_layers = nn.ModuleList(self.pyramid_pool_layers)
+
+    def forward(self, x):
+        x_size = x.size()
+        out = [x]
+        for layer in self.pyramid_pool_layers:
+            out.append(F.interpolate(layer(x), x_size[2:], mode='bilinear', align_corners=True))
+        return torch.cat(out, 1)
+        
+        
+class PSPNet(nn.Module):
+    def __init__(self, in_channels, num_classes, use_aux=False):
+        super(PSPNet, self).__init__()
+        self.in_channels = in_channels
+        self.num_classes = num_classes
+                
+        # backbone layers
+        backbone = resnet50(pretrained=True, replace_stride_with_dilation=[False, True, True])        
+        self.initial = nn.Sequential(*list(backbone.children())[:4])
+        self.layer1 = backbone.layer1
+        self.layer2 = backbone.layer2
+        self.layer3 = backbone.layer3
+        self.layer4 = backbone.layer4
+        
+        # Pyramid pooling module components
+        ppm_in_channels = int(backbone.fc.in_features)
+        self.ppm = pyramid_pooling_module(in_channels=ppm_in_channels, 
+                                     out_channels=512, bin_sizes=[1,2,3,6])
+        
+        # classifier head
+        self.cls = nn.Sequential(
+            nn.Conv2d(ppm_in_channels * 2, 512, kernel_size=3, padding=1, bias=False),
+            nn.BatchNorm2d(512),
+            nn.ReLU(inplace=True),
+            nn.Dropout2d(p=0.1),
+            nn.Conv2d(512,  self.num_classes, kernel_size=1)            
+        )
+        
+        # main branch is composed of PPM + Classifier
+        self.main_branch = nn.Sequential(self.ppm, self.cls)
+        
+        # Define Auxilary branch if specified
+        self.use_aux = False
+        if(self.training and use_aux):
+            self.use_aux = True
+            self.aux_branch = nn.Sequential(
+                nn.Conv2d( int(ppm_in_channels / 2) , 256, kernel_size=3, padding=1, bias=False),
+                nn.BatchNorm2d(256),
+                nn.ReLU(inplace=True),
+                nn.Dropout2d(p=0.1),
+                nn.Conv2d(256, self.num_classes, kernel_size=1)
+            )
+        
+        
+    def forward(self, x):
```

Comment:
I guess it could be nice if you could add some type hints to represent what we expect to have as an input and what do we expect to return- the [reference](https://docs.python.org/3/library/typing.html).
This is a very gentle way to communicate of the datatypes.

PS. The ultimate level is to introduce [Pydantic](https://docs.pydantic.dev/latest/) or the [Dataclasses](https://realpython.com/python-data-classes/) but I know that the most of the data sciencists ignores it ;)

Reply:
Ok, I'll add some information about input and output.